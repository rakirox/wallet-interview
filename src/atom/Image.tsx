import React, {useState} from 'react';
import {StyleSheet, Image as BaseImage, ImageProps} from 'react-native';

export function Image(tProps: ImageProps): JSX.Element {
  const [loading, setLoading] = useState(true);
  return (
    <>
      <BaseImage
        {...tProps}
        style={[styles.image, tProps.style]}
        onLoadEnd={() => setLoading(false)}
      />
      ;
    </>
  );
}

const styles = StyleSheet.create({
  image: {},
});
