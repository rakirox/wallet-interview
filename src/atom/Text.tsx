import React from 'react';
import {StyleSheet, Text as BaseText, TextProps} from 'react-native';

export function Text(tProps: TextProps): JSX.Element {
  return <BaseText {...tProps} style={[styles.text, tProps.style]} />;
}

const styles = StyleSheet.create({
  text: {
    fontFamily: 'avenir',
  },
});
