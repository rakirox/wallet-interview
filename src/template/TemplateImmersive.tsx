import React, {PropsWithChildren} from 'react';
import {SafeAreaView, StyleSheet, View} from 'react-native';

type TemplateProps = PropsWithChildren<{
  mainContent: JSX.Element;
  headerContent: JSX.Element;
  footContent: JSX.Element;
}>;

export function TemplateInmersive({
  headerContent,
  footContent,
  mainContent,
}: TemplateProps): JSX.Element {
  return (
    <View style={styles.flex}>
      <View style={styles.fakeBG} />
      <SafeAreaView style={styles.flex}>
        <View style={styles.headerContainer}>{headerContent}</View>
        <View style={styles.container}>
          <View style={styles.mainContent}>{mainContent}</View>
          <View style={styles.footContent}>{footContent}</View>
        </View>
      </SafeAreaView>
    </View>
  );
}

const styles = StyleSheet.create({
  fakeBG: {
    height: '50%',
    width: '100%',
    backgroundColor: '#CFD6FF',
    position: 'absolute',
    top: 0,
  },
  flex: {
    flex: 1,
  },
  footContent: {
    paddingHorizontal: 20,
  },
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
  headerContainer: {
    paddingVertical: 22,
    paddingHorizontal: 20,
  },
  mainContent: {
    backgroundColor: 'white',
    alignItems: 'center',
    marginBottom: 20,
    flex: 1,
  },
});
