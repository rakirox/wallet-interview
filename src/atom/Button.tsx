import React from 'react';
import {
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import {Text} from './Text';

type Props = {
  text: string;
  onPress: () => void;
  textStyle?: TextStyle;
} & TouchableOpacityProps;

export function Button({
  text,
  onPress,
  style,
  textStyle,
  ...tProps
}: Props): JSX.Element {
  return (
    <TouchableOpacity
      {...tProps}
      style={[styles.button, style]}
      onPress={onPress}>
      <Text style={[styles.text, textStyle, textStyle]}>{text}</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#334FFA',
    borderRadius: 10,
  },
  text: {
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
    color: 'white',
    alignSelf: 'center',
    paddingVertical: 14,
  },
});
