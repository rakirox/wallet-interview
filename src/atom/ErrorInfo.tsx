import React, {PropsWithChildren} from 'react';
import {View, StyleSheet} from 'react-native';
import {Text} from './Text';

type Props = PropsWithChildren<{
  text: string;
}>;

export function ErrorInfo({text}: Props) {
  return (
    <View style={styles.container}>
      <Text style={styles.errorText}>Error encontrado: {text}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
  errorText: {
    color: 'red',
  },
});
