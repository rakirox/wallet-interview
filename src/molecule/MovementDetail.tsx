import dayjs from 'dayjs';
import React, {PropsWithChildren} from 'react';
import {StyleSheet, View} from 'react-native';
import {Image} from '../atom/Image';
import {Text} from '../atom/Text';
import {Movement} from '../types/Movement';

type Props = PropsWithChildren<{
  movement: Movement;
}>;

export function MovementDetail({movement}: Props): JSX.Element {
  return (
    <View style={styles.container}>
      <View style={styles.shadow}>
        <View style={styles.imageContainer}>
          <Image source={{uri: movement.image}} style={styles.image} />
        </View>
      </View>
      <View style={styles.texts}>
        <Text style={styles.detailText}>Detalles del producto</Text>
        <Text style={styles.purchasedText}>{`Comprado el ${dayjs(
          new Date(movement.createdAt),
        )
          .locale('es')
          .format('DD [de] MMMM, YYYY')}`}</Text>
        <Text style={styles.withText}>Con esta compra acomulaste:</Text>
        <Text style={styles.pointsText}>{movement.points} puntos</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  pointsText: {
    fontSize: 24,
    fontWeight: '800',
  },
  withText: {
    fontSize: 14,
    fontWeight: '800',
    color: '#9B9898',
    marginBottom: 32,
  },
  purchasedText: {
    fontWeight: '800',
    fontSize: 16,
    marginBottom: 20,
  },
  detailText: {
    fontWeight: '800',
    fontSize: 14,
    color: '#9B9898',
    marginBottom: 19,
  },
  texts: {
    paddingHorizontal: 20,
    paddingVertical: 5,
  },
  container: {
    width: '100%',
  },
  shadow: {
    shadowColor: 'rgba(0, 0, 0, 0.5)',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowRadius: 3.84,
    shadowOpacity: 1,
    elevation: 6,
  },
  imageContainer: {
    borderRadius: 20,
    margin: 20,
    overflow: 'hidden',
    height: 350,
  },
  image: {
    flex: 1,
  },
});
