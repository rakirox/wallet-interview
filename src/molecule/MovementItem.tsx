import React, {PropsWithChildren} from 'react';
import {View, StyleSheet, Image, TouchableOpacity} from 'react-native';
import dayjs from 'dayjs';
import {Text} from '../atom/Text';
import {Movement} from '../types/Movement';

type Props = PropsWithChildren<{
  movement: Movement;
  onPress: () => void;
}>;

export function MovmentItem({movement, onPress}: Props): JSX.Element {
  const label = movement.isRedemption ? (
    <Text style={styles.negative}>-</Text>
  ) : (
    <Text style={styles.positive}>+</Text>
  );
  return (
    <TouchableOpacity style={styles.container} onPress={onPress}>
      <View style={styles.imageContainer}>
        <Image source={{uri: movement.image}} style={styles.image} />
      </View>
      <View style={styles.textContainer}>
        <Text style={styles.textProduct}>{movement.product}</Text>
        <Text style={styles.textPCreatedAt}>
          {dayjs(new Date(movement.createdAt))
            .locale('es')
            .format('DD [de] MMMM, YYYY')}
        </Text>
      </View>

      <Text style={styles.textPoints}>
        {label}
        {movement.points}
      </Text>

      <View style={styles.arrowContainer}>
        <Image
          source={require('./../assets/images/subtract.png')}
          style={styles.arrow}
        />
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  arrowContainer: {
    // flex: 1,
  },
  arrow: {
    width: 10,
    height: 10,
  },
  negative: {
    color: '#FF0000',
    fontStyle: 'normal',
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
  },
  positive: {
    color: '#00B833',
    fontStyle: 'normal',
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
  },
  container: {
    paddingBottom: 20,
    paddingHorizontal: 20,
    flexDirection: 'row',
    alignItems: 'center',
  },
  textContainer: {
    flex: 1,
  },
  textProduct: {
    fontWeight: '800',
    fontSize: 14,
    marginBottom: 7,
  },
  textPoints: {
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
    marginRight: 23,
  },
  textPCreatedAt: {
    fontWeight: '400',
    fontSize: 12,
  },
  image: {flex: 1},
  imageContainer: {
    borderRadius: 10,
    overflow: 'hidden',
    width: 55,
    height: 55,
    marginRight: 11,
  },
});
