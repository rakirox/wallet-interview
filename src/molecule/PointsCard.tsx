import React, {PropsWithChildren} from 'react';
import {ActivityIndicator, StyleSheet, View} from 'react-native';
import {Text} from '../atom/Text';

type Props = PropsWithChildren<{
  title: string;
  points: number;
  loading: boolean;
}>;

export function PointsCard({title, points, loading}: Props): JSX.Element {
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      {loading ? (
        <ActivityIndicator style={styles.activityIndicator} color="white" />
      ) : (
        <Text style={styles.points}>
          {`${points.toLocaleString('mxn', {
            minimumFractionDigits: 2,
          })} pts`}
        </Text>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  activityIndicator: {
    alignSelf: 'center',
    marginVertical: 30,
  },
  container: {
    backgroundColor: '#334FFA',
    padding: 20,
    shadowColor: 'rgba(0, 0, 0, 0.5)',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowRadius: 3.84,
    shadowOpacity: 1,
    elevation: 6,
    borderRadius: 20,
    minWidth: 280,
  },
  title: {
    fontWeight: '800',
    fontSize: 16,
    lineHeight: 22,
    color: '#ffffff',
    marginBottom: 7,
  },
  points: {
    fontWeight: '800',
    fontSize: 32,
    lineHeight: 44,
    color: '#ffffff',
    marginBottom: 50,
    alignSelf: 'center',
  },
});
