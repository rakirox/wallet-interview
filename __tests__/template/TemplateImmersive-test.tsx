import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {TemplateInmersive} from '../../src/template/TemplateImmersive';
import {movements} from '../mock.json';
import {Text} from '../../src/atom/Text';
import {MovementDetail} from '../../src/molecule/MovementDetail';
import {Button} from '../../src/atom/Button';

describe('TemplateInmersive', () => {
  it('render correctly', () => {
    expect(
      renderer.create(
        <TemplateInmersive
          headerContent={<Text>{movements[0].product}</Text>}
          mainContent={<MovementDetail movement={movements[0]} />}
          footContent={<Button text="Aceptar" onPress={jest.fn()} />}
        />,
      ),
    ).toMatchSnapshot();
  });
});
