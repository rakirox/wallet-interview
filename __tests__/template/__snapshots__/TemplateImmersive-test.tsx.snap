// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateInmersive render correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#CFD6FF",
        "height": "50%",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  />
  <RCTSafeAreaView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "paddingHorizontal": 20,
          "paddingVertical": 22,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "avenir",
            },
            undefined,
          ]
        }
      >
        Fantastic Granite Bacon
      </Text>
    </View>
    <View
      style={
        {
          "backgroundColor": "white",
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "white",
            "flex": 1,
            "marginBottom": 20,
          }
        }
      >
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "elevation": 6,
                "shadowColor": "rgba(0, 0, 0, 0.5)",
                "shadowOffset": {
                  "height": 4,
                  "width": 0,
                },
                "shadowOpacity": 1,
                "shadowRadius": 3.84,
              }
            }
          >
            <View
              style={
                {
                  "borderRadius": 20,
                  "height": 350,
                  "margin": 20,
                  "overflow": "hidden",
                }
              }
            >
              <Image
                onLoadEnd={[Function]}
                source={
                  {
                    "uri": "https://loremflickr.com/640/480/technics",
                  }
                }
                style={
                  [
                    {},
                    {
                      "flex": 1,
                    },
                  ]
                }
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              >
                <ActivityIndicator
                  color="rgba(0,0,0,0.5)"
                />
              </View>
            </View>
          </View>
          <View
            style={
              {
                "paddingHorizontal": 20,
                "paddingVertical": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "color": "#9B9898",
                    "fontSize": 14,
                    "fontWeight": "800",
                    "marginBottom": 19,
                  },
                ]
              }
            >
              Detalles del producto
            </Text>
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "fontSize": 16,
                    "fontWeight": "800",
                    "marginBottom": 20,
                  },
                ]
              }
            >
              Comprado el 09 de December, 2022
            </Text>
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "color": "#9B9898",
                    "fontSize": 14,
                    "fontWeight": "800",
                    "marginBottom": 32,
                  },
                ]
              }
            >
              Con esta compra acomulaste:
            </Text>
            <Text
              style={
                [
                  {
                    "fontFamily": "avenir",
                  },
                  {
                    "fontSize": 24,
                    "fontWeight": "800",
                  },
                ]
              }
            >
              42416
               puntos
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "paddingHorizontal": 20,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#334FFA",
              "borderRadius": 10,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "avenir",
                },
                [
                  {
                    "alignSelf": "center",
                    "color": "white",
                    "fontSize": 16,
                    "fontWeight": "800",
                    "lineHeight": 22,
                    "paddingVertical": 14,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Aceptar
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
